cluster/clustered_data.go

```diff
@@ -0,0 +1,199 @@
+implementation of conditional PUTs requires us to create a new object on each update. If the object is large this
+results in a lot of objects and a lot of storage under load.
+
+Before the call to `StoreState` returns a no-op update on `StatePersistor` occurs to read the latest epoch. If this does
+not match the epoch the state was stored with thew call returns an error.
+
+The key that we store the state under includes the epoch in it, e.g. `my-data-key-prefix-0000000234`. When data is
+loaded, the key with the highest epoch prefix is loaded.
+*/
+type ClusteredData struct {
+	lock           sync.Mutex
+	dataBucketName string
+	dataKeyPrefix  string
+	stateMachine   *StateUpdator
+	objStoreClient objstore.Client
+	epoch          uint64
+	opts           ClusteredDataOpts
+	readyState     clusteredDataState
+	stopping       atomic.Bool
+}
+
+type clusteredDataState int
+
+const (
+	clusteredDataStateReady = iota
+	clusteredDataStateLoaded
+	clusteredDataStateStopped
+)
+
+func NewClusteredData(stateMachineBucketName string, stateMachineKeyPrefix string, dataBucketName string,
+	dataKeyPrefix string, objStoreClient objstore.Client, opts ClusteredDataOpts) *ClusteredData {
+	return &ClusteredData{
+		objStoreClient: objStoreClient,
+		dataBucketName: dataBucketName,
+		dataKeyPrefix:  dataKeyPrefix,
+		stateMachine: NewStateUpdator(stateMachineBucketName, stateMachineKeyPrefix, objStoreClient,
+			StateUpdatorOpts{}),
+		opts:       opts,
+		readyState: clusteredDataStateReady,
+	}
+}
+
+type ClusteredDataOpts struct {
+	AvailabilityRetryInterval time.Duration
+	ObjStoreCallTimeout       time.Duration
+}
+
+func (mo *ClusteredDataOpts) setDefaults() {
+	if mo.AvailabilityRetryInterval == 0 {
+		mo.AvailabilityRetryInterval = DefaultAvailabilityRetryInterval
+	}
+}
+
+func (m *ClusteredData) LoadData() ([]byte, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.readyState == clusteredDataStateStopped {
+		return nil, errors.New("stopped")
+	}
+	// Atomically increment the epoch
+	buff, err := m.stateMachine.Update(func(state []byte) ([]byte, error) {
+		epoch := buffToEpoch(state)
+		newState := make([]byte, 8)
+		binary.BigEndian.PutUint64(newState, epoch+1)
+		return newState, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	m.epoch = buffToEpoch(buff)
+	prevEpoch := m.epoch - 1
+	// Now we try and load the key with the highest epoch - the key might be lower than prevEpoch as no data might
+	// have been stored in previous epoch
+	var data []byte
+	for {
+		dataKey := m.createDataKey(prevEpoch)
+		data, err = m.getWithRetry(dataKey)
+		if err != nil {
+			return nil, err
+		}
+		if len(data) == 0 {
+			// no key found - try with next lower epoch
+			if prevEpoch == 0 {
+				// No data to load
+				break
+			}
+			prevEpoch--
+		} else {
+			break
+		}
+	}
+	m.readyState = clusteredDataStateLoaded
+	return data, nil
+}
+
+func (m *ClusteredData) StoreData(data []byte) (bool, error) {
```

Comment:
Typical usage would be: one call of LoadData() to get the current data when the controller becomes leader, then it can call StoreData() as many times as it likes. 
If another controller becomes leader during this period, then the next call to StoreData() will fail.
Epoch is only update when LoadData() is called, i.e. when controller becomes leader.

Reply:
we should document this - it wasn't straightforward to me.
Perhaps `LoadData` should be called `AcquireData` ? Something that implies you're the only one getting it